text/0000-default-field-values.md

```diff
@@ -0,0 +1,1802 @@
+from constants.
+
+## `#[derive(Default)]`
+
+When generating an implementation of `Default` for a `struct` named `$s` on
+which `#[derive(Default)]` has been attached, the compiler will omit all fields
+which have default values provided in the `struct`. The the associated function
+`default` shall then be defined as (where `$f_i` denotes the `i`-th field of
+`$s`):
+
+```rust
+fn default() -> Self {
+    $s { $f_i: Default::default(), .. }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The usual drawback of increasing the complexity of the language applies.
+However, the degree to which complexity is increased is not substantial.
+
+In particular, the syntax `Foo { .. }` mirrors the identical and already
+existing pattern syntax. This makes the addition of `Foo { .. }` at worst
+low-cost and potentially cost-free.
+
+It is true that there are cases where `Foo { ..Default::default() }` will be
+allowed where `Foo { .. }` won't be, and vice-versa.
+
+This new syntax is more ergonomic to use, but it requires specifying a default
+value for every field which can be much less ergonomic than using
+`#[derive(Default)]` on your type. The following two are almost equivalent, and
+the more fields there are, the more the verbosity is increased:
+
+```rust
+#[derive(Default)]
+struct S {
+    foo: Option<String>,
+    bar: Option<String>,
+}
+```
+
+```rust
+struct S {
+    foo: Option<String> = None,
+    bar: Option<String> = None,
+}
+```
+
+This can become relevant when an API author wants to push users towards the new
+syntax because `..` is shorter than `..Default::default()`, or when some fields
+with types that `impl Default` are optional, but `#[derive(Default)]` can't be
+used because some fields are mandatory.
+
+The main complexity comes instead from introducing `field: Type = expr`.
+However, as seen in the [prior-art], there are several widely-used languages
+that have a notion of field / property / instance-variable defaults.
+Therefore, the addition is intuitive and thus the cost is seen as limited. As
+an implementation detail, `rustc` *already* parses `field: Type = expr`
+purely to provide an appropriate diagnostic error:
+
+```
+error: default values on `struct` fields aren't supported
+ --> src/lib.rs:2:28
+  |
+2 |     pub alpha: &'static str = "",
+  |                            ^^^^^ help: remove this unsupported default value
+```
+
+An issue arises when considering `const` patterns. A pattern `Foo { .. }` can
+match more things than just the expression `Foo { .. }`, because the pattern
+matches any value of the unmentioned fields, but the expression sets them to a
+particular value. This means that, with the unstable `inline_const_pat`, the arm
+`const { Foo { .. } } =>` matches less than the arm `Foo { .. } =>` (assuming a
+type like `struct Foo { a: i32 = 1 }`). A way to mitigate this might be to use
+an alternative syntax, like `...` or `..kw#default`.
+
```

Comment:
Maybe I missed how it's prevented, but it seems like a drawback not called out here is that `Foo { .. }` and `Foo::default()` could mean different things with the RFC as proposed, since the default impl can be manual and differ significantly in terms of the values it provides.

Should we perhaps forbid mixing manual Default impls and the default field value syntax, or otherwise at least suggest something here?

Reply:
Oversight on my part (that was brought up in my original thread).

I feel like it would be best to *lint* against that (maybe even deny-by-default), but otherwise allow it. It is a foot-gun, but there might be good reasons for an API to be designed that way (even if I can't picture them now). We can also disallow that today and allow it in the future.

Reply:
We can always make the lint check for that, triggering only if the manual `impl` is identical to what we would have derived.

Reply:
The above is now mentioned in Future possibilities